File: src/bin/0199-binary-tree-right-side-view.py
# @leetup=info id=199 lang=python3 slug=binary-tree-right-side-view



from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @leetup=code


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> list[int]:
        res = []

        def dfs(n, level):
            if not n:
                return
            if len(res) == level:
                res.append(n.val)
            dfs(n.right, level + 1)
            dfs(n.left, level + 1)

        dfs(root, 0)
        return res


# @leetup=code

File: src/bin/0199-binary-tree-right-side-view.rs
// @leetup=info id=199 lang=rust slug=binary-tree-right-side-view


// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @leetup=code
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        fn dfs(root: Option<Rc<RefCell<TreeNode>>>, level: usize, res: &mut Vec<i32>) {
            if let Some(root) = root {
                if res.len() == level {
                    res.push(root.borrow().val);
                }
                dfs(root.borrow().right.clone(), level + 1, res);
                dfs(root.borrow().left.clone(), level + 1, res);
            }
        }

        let mut res = Vec::new();
        dfs(root, 0, &mut res);
        res
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::right_side_view;
}
